soroban-env-host/src/budget.rs

```diff
@@ -118,23 +123,38 @@ impl CostType {
 
 #[derive(Default, Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
 pub struct CostModel {
-    pub const_param: u64,
-    pub lin_param: u64,
+    pub(crate) const_param: u64,
+    pub(crate) lin_param: u64,
+    pub(crate) iterations: u64,
 }
 
 impl CostModel {
-    pub fn evaluate(&self, input: u64) -> u64 {
-        let mut res = self.const_param;
-        if self.lin_param != 0 {
-            res = res.saturating_add(self.lin_param.saturating_mul(input));
-        }
-        res
+    /// Evaluates the linear model `f(input) = const_param + lin_param * input`.
+    /// If the input is `None`, then the input is ignored, resulting in a constant
+    /// model.
+    pub fn evaluate(&self, iterations: u64, input: Option<u64>) -> u64 {
+        if iterations == 0 {
+            return 0;
+        };
+
+        let eval = match input {
+            Some(input) => {
+                let mut res = self.const_param;
+                if self.lin_param != 0 {
+                    res = res.saturating_add(self.lin_param.saturating_mul(input));
+                }
+                res
+            }
+            None => self.const_param,
+        };
+        iterations.saturating_mul(eval)
```

Comment:
So here you're using the iteration count _passed in_ and ignoring the one stored in the model. Which I think is closer to making sense for case 3 above, but .. again I'm not 100% sure the cases that use it outside of calibration really need it.